text/0020-unique-record-keys.md

```diff
@@ -0,0 +1,103 @@
+If duplicate keys were prohibited, the order of keys wouldn't matter (but see "Unresolved questions" below).
+
+A tertiary motivation for this proposal is that it would bring the Cedar language in line with our desired behavior on duplicate keys in other Cedar API surfaces. See [cedar#159](https://github.com/cedar-policy/cedar/issues/159).
+
+## Detailed design
+
+We'd do something very like [cedar#169](https://github.com/cedar-policy/cedar/pull/169), except that instead of implicitly dropping all but the last value for each key (in `.collect()`), we'd be careful to emit an error when constructing an `Expr` if the record has duplicate keys.
+
+In the case of record literals, the error could be caught in the CST->AST step, resulting in the user seeing a parse error rather than an evaluation error.
+In the case of `context`, the error would need to be caught when constructing the `Context` object.
+In other cases, e.g. if evaluating a Cedar subexpression produces a record with duplicate keys, the user would see an evaluation error. (But see "Unresolved questions" below.)
+
+I don't believe this change would have a significant effect on validation or width/depth subtyping of records, but please correct me if I'm wrong.
+
+## Drawbacks
+
+This is a breaking change, not just to the Cedar API surface, but to the Cedar language (policy syntax) itself.
+That's a significant drawback not to be taken lightly.
+In this case, it's mitigated somewhat by the fact that we're turning previously-valid policies into errors, not changing them into a silently different behavior.
+It's also mitigated by the fact that most Cedar users are probably (citation needed) not writing policies in the class that will have a behavior change -- i.e., policies with duplicate keys in record literals.
+
+## Alternatives
+
+The main alternative is doing nothing, which of course has the reverse pros/cons of the proposal as explained in "Motivation" and "Drawbacks".
+
+Another alternative, which would address the first motivation and not the second, would be to change the EST format so that it supports record literals with duplicate keys.
+I believe this would require a breaking change to the EST format, and record literals would not be expressible as JSON maps, which is unfortunate because I believe that's the most natural/ergonomic.
+(Devil's advocate: we could maybe find a way to support JSON maps in the common case with some escape syntax for expressing record literals that have duplicate keys. If we want to take this alternative seriously, I can investigate this.)
+
+## Unresolved questions
+
+### Evaluation errors for duplicate keys
+Is it actually possible in today's Cedar, to have duplicate keys in a record value other than a record literal or `context`? I.e., to have duplicate keys in a record value produced by evaluating some subexpression (other than a record literal or the `context` variable)?
```

Comment:
This is not possible today because record literals are the only way to construct a record value. All values that are not literals are coming from the input (context or entities).

Reply:
your comment specifically mentions a case (entities) that's not covered here.  Worth investigating duplicate keys in records in entity attributes

Reply:
Revised RFC text per this discussion